drivers/flash/nrf_qspi_nor.c

```diff
@@ -262,13 +276,23 @@ static inline void qspi_wait_for_completion(const struct device *dev,
 	struct qspi_nor_data *dev_data = get_dev_data(dev);
 
 	if (res == NRFX_SUCCESS) {
+#ifdef CONFIG_MULTITHREADING
 		k_sem_take(&dev_data->sync, K_FOREVER);
+#else /* CONFIG_MULTITHREADING */
+		while (!atomic_cas(&dev_data->ready, 1, 0)) {
+			/* spin */
```

Comment:
Is the variable going to become ready in an IRQ context? If so, I wonder if we could replace the spin with cpu-idling. Otherwise, it should be as it is now

Reply:
It is, and I have no idea how to idle the cpu and still detect completion without explicitly using `SEV/WFE` instructions which seems too low-level.  Is there existing arch API for this?

Reply:
Yes, there is k_cpu_idle() which uses WFI in Cortex-M.

Reply:
And is that guaranteed reliable?  Normally I do `SEV;WFE` at a point where I know it's before the signal, to avoid a race condition of idling just after the flag was set.

Reply:
OK, it seems to work, gets rid of the ready flag.  Will update.

Reply:
what is the guaranty that thread will be resumed only on a qspi complete operation event.

Reply:
That this is running the only "thread".  It also means there can't be any other pending interrupt activities that might wake it up instead.  Which is reasonable to expect with a threadless system, but does make `k_timer` unsupportable.  (I think `k_sleep()` can be made to work, though.)

If we go this way we won't be able to overlap operations that signal completion with an interrupt, though.  That's probably a limitation we'll have to live with; or we revert to the spin-wait solution and never idle; or to idle we'd need an event recording solution like [this](https://pabigot.github.io/nrfcxx/classnrfcxx_1_1event__set.html#details).

Reply:
@pabigot do you need to lock interrupts before calling k_cpu_idle()? That's because k_cpu_idle() in Cortex-m uses WFI, not WFE. what will happen if the ISR fires before you manage to do CPSID inside arch_cpu_idle?

Reply:
Yeah, on further thought none of these solutions are quite what I want.  I'll rework this later this morning when I can test multiple alternatives.